Panic on `toMap` in a lambda when its argument is a parameter

I got the same error in pydhall (at type check time, though, not at eval)

```shell
$ go run cmd/main.go <<<('let l = \(input : { a : Natural }) -> toMap input in l { a = 1 }')
panic: unknown term type<nil>

goroutine 1 [running]:
github.com/philandstuff/dhall-golang/v4/core.evalWith(0x0, 0x0, 0xc00027b4d0, 0x7ff240, 0xc00027e920)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/eval.go:615 +0x76a9
github.com/philandstuff/dhall-golang/v4/core.evalWith(0x7ff680, 0xc00027e7c0, 0xc00027b4d0, 0x5, 0xc00012cb88)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/eval.go:421 +0x4862
github.com/philandstuff/dhall-golang/v4/core.evalWith.func1(0x7ff240, 0xc00027e920, 0x7ff240, 0xc00027e920)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/eval.go:103 +0x2c9
github.com/philandstuff/dhall-golang/v4/core.lambda.Call(...)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/ast.go:196
github.com/philandstuff/dhall-golang/v4/core.quoteWith(0xc00019db28, 0xc00027af00, 0x7fef40, 0xc00027b4a0, 0xc00027b4a0, 0x0)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/quote.go:149 +0x3dea
github.com/philandstuff/dhall-golang/v4/core.Quote(...)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/quote.go:11
github.com/philandstuff/dhall-golang/v4/core.typeWith(0xc00019fd28, 0x7ff4e0, 0xc00027b020, 0x0, 0x0, 0x7ff4e0, 0xc00027b020)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/typecheck.go:265 +0xa962
github.com/philandstuff/dhall-golang/v4/core.TypeOf(...)
	/home/lisael/projects/pydhall/ext/dhall-golang/core/typecheck.go:60
main.main()
	/home/lisael/projects/pydhall/ext/dhall-golang/cmd/main.go:24 +0x20e
exit status 2
```

This is because term.ToMap.Type is not set. If the expression is annotated, everything's fine : 
```sh
$ go run cmd/main.go <<<('let l = \(input : { a : Natural }) -> toMap input : ( List { mapKey : Text, mapValue : Natural }) in l { a = 1 }')
{map[mapKey:Text mapValue:Natural]}
[map[mapKey:a mapValue:1]]
decoded as [{ mapKey = "a", mapValue = 1}]
```
... as expected.

I'm not sure where to inject the result of the type inference.

--- a/core/eval.go
+++ b/core/eval.go
@@ -416,10 +416,11 @@ func evalWith(t term.Term, e env) Value {
 			}
 			return result
 		}
-		return toMap{
-			Record: record,
-			Type:   evalWith(t.Type, e),
+		toMapVal := toMap{Record: recordVal}
+		if t.Type != nil {
+			toMapVal.Type = evalWith(t.Type, e)
 		}
+		return toMapVal
 	case term.Field:
 		record := evalWith(t.Record, e)
 		for { // simplifications

--- a/core/eval_test.go
+++ b/core/eval_test.go
@@ -38,4 +38,14 @@ var _ = Describe("Eval", func() {
 				To(Equal(Type))
 		})
 	})
+	Describe("toMap", func() {
+		It("Evaluates with missing type and abstract value", func() {
+			Expect(Eval(term.ToMap{
+				Record: term.Var{Name: "x"},
+			})).
+				To(Equal(toMap{
+					Record: freeVar{Name: "x"},
+				}))
+		})
+	})
 })

